java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQ.java

```diff
@@ -0,0 +1,173 @@
+  private String kafkaStartingOffsets;
+  private Long kafkaAwaitTerminationTimeout;
+  private String failOnDataLoss;
+  private String bigQueryDataset;
+  private String bigQueryTable;
+  private String tempGcsBucket;
+  private String streamOutputMode;
+
+  public KafkaToBQ() {
+    projectId = getProperties().getProperty(PROJECT_ID_PROP);
+    kafkaBootstrapServers = getProperties().getProperty(KAFKA_BQ_BOOTSTRAP_SERVERS);
+    kafkaTopic = getProperties().getProperty(KAFKA_BQ_TOPIC);
+    checkpointLocation = getProperties().getProperty(KAFKA_BQ_CHECKPOINT_LOCATION);
+    kafkaStartingOffsets = getProperties().getProperty(KAFKA_BQ_STARTING_OFFSET);
+    failOnDataLoss = getProperties().getProperty(KAFKA_BQ_FAIL_ON_DATA_LOSS);
+    bigQueryDataset = getProperties().getProperty(KAFKA_BQ_DATASET);
+    bigQueryTable = getProperties().getProperty(KAFKA_BQ_TABLE);
+    tempGcsBucket = getProperties().getProperty(KAFKA_BQ_TEMP_GCS_BUCKET);
+    kafkaAwaitTerminationTimeout =
+        Long.valueOf(getProperties().getProperty(KAFKA_BQ_AWAIT_TERMINATION_TIMEOUT));
+    streamOutputMode = getProperties().getProperty(KAFKA_BQ_STREAM_OUTPUT_MODE);
+  }
+
+  @Override
+  public void runTemplate() {
+    if (StringUtils.isAllBlank(checkpointLocation)
+        || StringUtils.isAllBlank(kafkaBootstrapServers)
+        || StringUtils.isAllBlank(kafkaTopic)
+        || StringUtils.isAllBlank(bigQueryDataset)
+        || StringUtils.isAllBlank(bigQueryTable)
+        || StringUtils.isAllBlank(projectId)
+        || StringUtils.isAllBlank(tempGcsBucket)) {
+      LOGGER.error(
+          "{},{},{},{},{},{},{} is required parameter. ",
+          PROJECT_ID_PROP,
+          KAFKA_BQ_CHECKPOINT_LOCATION,
+          KAFKA_BQ_BOOTSTRAP_SERVERS,
+          KAFKA_BQ_TOPIC,
+          KAFKA_BQ_DATASET,
+          KAFKA_BQ_TABLE,
+          KAFKA_BQ_TEMP_GCS_BUCKET);
+      throw new IllegalArgumentException(
+          "Required parameters for KafkaToBQ not passed. "
+              + "Set mandatory parameter for KafkaToBQ template "
+              + "in resources/conf/template.properties file.");
+    }
+
+    SparkSession spark = null;
+    LOGGER.info(
+        "Starting Kafka to BQ spark job with following parameters:"
+            + "1. {}:{} "
+            + "2. {}:{} "
+            + "3. {}:{} "
+            + "4. {},{} "
+            + "5. {},{} "
+            + "6. {},{} "
+            + "7. {},{} "
+            + "8. {},{} ",
+        KAFKA_BQ_CHECKPOINT_LOCATION,
+        checkpointLocation,
+        KAFKA_BQ_BOOTSTRAP_SERVERS,
+        kafkaBootstrapServers,
+        KAFKA_BQ_TOPIC,
+        kafkaTopic,
+        KAFKA_BQ_STARTING_OFFSET,
+        kafkaStartingOffsets,
+        KAFKA_BQ_AWAIT_TERMINATION_TIMEOUT,
+        kafkaAwaitTerminationTimeout,
+        KAFKA_BQ_DATASET,
+        bigQueryDataset,
+        KAFKA_BQ_TABLE,
+        bigQueryTable,
+        KAFKA_BQ_TEMP_GCS_BUCKET,
+        tempGcsBucket);
+
+    try {
```

Comment:
no try catch please

Reply:
@shashank-google The template is designed to cater the capabilities of both a batch load and a streaming load.

Unfortunately, for streaming applications, which can raise org.apache.spark.sql.streaming.StreamingQueryException. It is necessary to either catch or declare this StreamingQueryException to be thrown.

Already tried avoiding try-catch for the same, getting: `org.apache.spark.sql.streaming.StreamingQueryException must be caught or declared to be thrown`


1. declared to be thrown: Would not work as it requires change in BaseTemplate to declare runTemplate throws an Exception
2. try-catch clause is the only option left out for these kind of templates (Kafka, PubSub, etc).

Reply:
If that is the case, it is not appropriate to use generic `catch (Throwable t)`
However it is better to go with (1) i.e. modify the BaseTemplate.runTemplate with `throws` clause in method signature.
Try-catch should "only" be used if you are writing a logic which can recover from failure "automatically".